src/python/createrepo_c/__init__.py

```diff
@@ -382,6 +386,364 @@ def __init__(self, primary_path, filelists_path, other_path, newpkgcb=None, warn
         _createrepo_c.PkgIterator.__init__(
             self, primary_path, filelists_path, other_path, newpkgcb, warningcb)
 
+
+class RepositoryReader:
+    """Parser for RPM metadata."""
+
+    def __init__(self):
+        """Initialize empty (use one of the alternate constructors)."""
+        self._primary_xml_path = None
+        self._filelists_xml_path = None
+        self._other_xml_path = None
+        self._updateinfo_path = None
+        self.repomd = None
+
+    @staticmethod
+    def from_path(path, warningcb=None):
+        """Construct a parser from an on-disk repository."""
+
+        if not warningcb:
+            def _warningcb(warning_type, message):
+                print("PARSER WARNING: %s" % message)
+                return True
+            warningcb = _warningcb
+
+        repomd_path = Path(path) / "repodata" / "repomd.xml"
+        if not repomd_path.exists():
+            raise FileNotFoundError("No repository found at the provided path.")
+
+        repomd = Repomd(str(repomd_path))
+        metadata_files = {record.type: record for record in repomd.records}
+        parser = RepositoryReader()
+        parser._primary_xml_path = os.path.join(path, metadata_files["primary"].location_href)
+        parser._filelists_xml_path = os.path.join(path, metadata_files["filelists"].location_href)
+        parser._other_xml_path = os.path.join(path, metadata_files["other"].location_href)
```

Comment:
Hmm, what about creating some structure for passing these metadata file paths? I see it on many places in the script and also in dnf the filelists are already optional, so user shouldn't be forced to fill it I guess.

Reply:
I figured that `parser.repomd` (the repomd file) is that data structure already

Reply:
I was rather thinking of using some single wrapper for all these *_path here, in `from_metadata_files`, in `PackageIterator`, like everywhere these paths are on the input/output. It's just an idea. But anyway, the filelists arg in `from_metadata_files` should be optional at least...

Reply:
It can't be made optional currently because the underlying C API doesn't make filelists optional.  It can, eventually, but it's not currently done that way.

Reply:
I am sorry, I need to look into it deeper again, but probably I'll have a chance yet in the next week.

Reply:
I've created a separate issue for this: https://github.com/rpm-software-management/createrepo_c/issues/423.